dask/array/_shuffle.py

```diff
@@ -10,15 +10,20 @@
 import numpy as np
 
 from dask import config
-from dask._task_spec import DataNode, List, Task, TaskRef
+from dask._task_spec import DataNode, GraphNode, List, Task, TaskRef
 from dask.array.chunk import getitem
 from dask.array.core import Array, unknown_chunk_message
 from dask.array.dispatch import concatenate_lookup, take_lookup
 from dask.base import tokenize
 from dask.highlevelgraph import HighLevelGraph
 
 
-def shuffle(x, indexer: list[list[int]], axis: int, chunks: Literal["auto"] = "auto"):
+def shuffle(
+    x: Array,
+    indexer: list[list[int]],
+    axis: int,
+    chunks: Literal["auto"] | tuple[tuple[int, ...], ...] = "auto",
```

Comment:
this may be easier to interpret as `chunks: dict[int, tuple[tuple[int, ...], ...]` where `axis not in chunks`?

Reply:
For example this fails:
```python
array = dask.array.ones(shape=(40, 8), chunks=(20, 1))
dask.array.shuffle(array, axis=1, indexer=[[0,0,0,0], [1,1,1,1], [2,2,2,2]], chunks=((20,20),))
```

Reply:
Alternatively if you accept `chunks` this way, then you effectively have `take` with  specified `chunks` don't you? 

Also the code [here](https://github.com/pydata/xarray/pull/9896/files#) combines `shuffle` and `reshape_blockwise` basically to get `.take(..., indices=nDarray, chunks=nDchunks)` IIUC.

Reply:
This is a good point. I like the dict API!

Reply:
take by itself doesn't guarantee that specific groups are ending up in the same chunk and I'd rather not guarantee that, that takes a lot of flexibility away in how we build stuff up. Eventually though, we are calling shuffle from take anyway, just with more flexible chunks. We can also work on the take interface to accept nested indexers if that would help somehow?

Reply:
> take by itself doesn't guarantee that specific groups are ending up in the same chunk and I'd rather not guarantee that, that takes a lot of flexibility away in how we build stuff up

I'm suggesting that it could make this guarantee IFF `chunks` are (optionally) provided.